File: quota.py
import time
import threading
from collections import defaultdict

class TenantQuota:
    def __init__(self, tenant_id, max_events_per_sec=0, max_storage_bytes=0, max_query_per_min=0, disabled=False):
        self.tenant_id = tenant_id
        self.max_events_per_sec = max_events_per_sec
        self.max_storage_bytes = max_storage_bytes
        self.max_query_per_min = max_query_per_min
        self.disabled = disabled

class TenantUsage:
    def __init__(self, tenant_id):
        self.tenant_id = tenant_id
        self.events_this_sec = 0
        self.events_this_day = 0
        self.storage_used = 0
        self.queries_this_min = 0
        self.last_event_time = 0
        self.last_query_time = 0

class QuotaManager:
    def __init__(self):
        self._lock = threading.RLock()
        self.quotas = {}
        self.usage = defaultdict(lambda: None)

    def set_quota(self, quota):
        with self._lock:
            self.quotas[quota.tenant_id] = quota

    def is_disabled(self, tenant_id):
        with self._lock:
            q = self.quotas.get(tenant_id)
            return q.disabled if q else False

    def check_event_rate(self, tenant_id):
        with self._lock:
            q = self.quotas.get(tenant_id)
            if not q or q.max_events_per_sec <= 0:
                return True
            if q.disabled:
                return False

            usage = self.usage.get(tenant_id)
            if usage is None:
                usage = TenantUsage(tenant_id)
                self.usage[tenant_id] = usage

            now = time.time()
            if now - usage.last_event_time >= 1:
                usage.events_this_sec = 0
            usage.events_this_sec += 1
            usage.last_event_time = now

            return usage.events_this_sec <= q.max_events_per_sec

    def check_storage(self, tenant_id, additional_bytes):
        with self._lock:
            q = self.quotas.get(tenant_id)
            if not q or q.max_storage_bytes <= 0:
                return True
            usage = self.usage.get(tenant_id)
            if not usage:
                return True
            return usage.storage_used + additional_bytes <= q.max_storage_bytes

    def add_storage_usage(self, tenant_id, bytes_used):
        with self._lock:
            usage = self.usage.get(tenant_id)
            if usage is None:
                usage = TenantUsage(tenant_id)
                self.usage[tenant_id] = usage
            usage.storage_used += bytes_used

File: services/shared/tenant/quota.go
package tenant

import (
    "fmt"
    "sync"
    "time"
)

// QuotaManager 租户配额管理器
type QuotaManager struct {
    mu       sync.RWMutex
    quotas   map[string]*TenantQuota // tenant_id -> quota
    usage    map[string]*TenantUsage // tenant_id -> usage
    // redis    RedisClient // optional, for distributed rate limiting
}

// TenantQuota 租户配额
type TenantQuota struct {
    TenantID        string
    MaxEventsPerSec int       // 每秒最大事件数
    MaxStorageBytes int64     // 最大存储字节数
    MaxQueryPerMin  int       // 每分钟最大查询数
    Disabled        bool      // 租户是否被禁用
    CreatedAt       time.Time
    UpdatedAt       time.Time
}

// TenantUsage 租户当前使用量
type TenantUsage struct {
    TenantID        string
    EventsThisSec   int       // 当前秒事件数
    EventsThisDay   int       // 当天事件数
    StorageUsed     int64     // 已用存储
    QueriesThisMin  int       // 当前分钟查询数
    LastEventTime   time.Time
    LastQueryTime   time.Time
}

// NewQuotaManager 创建配额管理器
func NewQuotaManager() *QuotaManager {
    return &QuotaManager{
        quotas: make(map[string]*TenantQuota),
        usage:  make(map[string]*TenantUsage),
    }
}

// SetQuota 设置租户配额
func (qm *QuotaManager) SetQuota(q *TenantQuota) {
    qm.mu.Lock()
    defer qm.mu.Unlock()
    q.UpdatedAt = time.Now()
    qm.quotas[q.TenantID] = q
}

// GetQuota 获取租户配额
func (qm *QuotaManager) GetQuota(tenantID string) *TenantQuota {
    qm.mu.RLock()
    defer qm.mu.RUnlock()
    return qm.quotas[tenantID]
}

// IsDisabled 检查租户是否被禁用
func (qm *QuotaManager) IsDisabled(tenantID string) bool {
    qm.mu.RLock()
    defer qm.mu.RUnlock()
    q, ok := qm.quotas[tenantID]
    if !ok {
        return false // 默认不禁用
    }
    return q.Disabled
}

// CheckEventRate 检查事件写入速率
func (qm *QuotaManager) CheckEventRate(tenantID string) error {
    qm.mu.Lock()
    defer qm.mu.Unlock()

    quota, ok := qm.quotas[tenantID]
    if !ok {
        return nil // 无配额限制
    }

    if quota.Disabled {
        return fmt.Errorf("tenant %s is disabled", tenantID)
    }

    usage, ok := qm.usage[tenantID]
    if !ok {
        usage = &TenantUsage{TenantID: tenantID}
        qm.usage[tenantID] = usage
    }

    now := time.Now()
    if now.Sub(usage.LastEventTime) >= time.Second {
        usage.EventsThisSec = 0
    }
    usage.EventsThisSec++
    usage.LastEventTime = now

    if quota.MaxEventsPerSec > 0 && usage.EventsThisSec > quota.MaxEventsPerSec {
        return fmt.Errorf("tenant %s rate limit exceeded: %d/%d events/sec",
            tenantID, usage.EventsThisSec, quota.MaxEventsPerSec)
    }

    return nil
}

// CheckQueryRate 检查查询速率
func (qm *QuotaManager) CheckQueryRate(tenantID string) error {
    qm.mu.Lock()
    defer qm.mu.Unlock()

    quota, ok := qm.quotas[tenantID]
    if !ok {
        return nil
    }

    if quota.Disabled {
        return fmt.Errorf("tenant %s is disabled", tenantID)
    }

    usage, ok := qm.usage[tenantID]
    if !ok {
        usage = &TenantUsage{TenantID: tenantID}
        qm.usage[tenantID] = usage
    }

    now := time.Now()
    if now.Sub(usage.LastQueryTime) >= time.Minute {
        usage.QueriesThisMin = 0
    }
    usage.QueriesThisMin++
    usage.LastQueryTime = now

    if quota.MaxQueryPerMin > 0 && usage.QueriesThisMin > quota.MaxQueryPerMin {
        return fmt.Errorf("tenant %s query limit exceeded: %d/%d queries/min",
            tenantID, usage.QueriesThisMin, quota.MaxQueryPerMin)
    }

    return nil
}

// CheckStorage 检查存储配额
func (qm *QuotaManager) CheckStorage(tenantID string, additionalBytes int64) error {
    qm.mu.RLock()
    defer qm.mu.RUnlock()

    quota, ok := qm.quotas[tenantID]
    if !ok {
        return nil
    }

    if quota.MaxStorageBytes <= 0 {
        return nil
    }

    usage, ok := qm.usage[tenantID]
    if !ok {
        return nil
    }

    if usage.StorageUsed+additionalBytes > quota.MaxStorageBytes {
        return fmt.Errorf("tenant %s storage limit exceeded: %d/%d bytes",
            tenantID, usage.StorageUsed+additionalBytes, quota.MaxStorageBytes)
    }

    return nil
}

// AddStorageUsage 增加存储使用量
func (qm *QuotaManager) AddStorageUsage(tenantID string, bytes int64) {
    qm.mu.Lock()
    defer qm.mu.Unlock()

    usage, ok := qm.usage[tenantID]
    if !ok {
        usage = &TenantUsage{TenantID: tenantID}
        qm.usage[tenantID] = usage
    }
    usage.StorageUsed += bytes
}
